tests/unit/check_terminal.c

```diff
@@ -0,0 +1,246 @@
+    ck_assert_int_eq(rv, OSD_OK);
+}
+
+void teardown_hostmod()
+{
+    osd_result rv;
+
+    // deactivate event sending
+    mock_host_controller_expect_reg_read(mock_hostmod_diaddr,
+                                         mock_dem_uart_diaddr,
+                                         OSD_REG_BASE_MOD_CS, 1);
+    mock_host_controller_expect_reg_write(mock_hostmod_diaddr,
+                                          mock_dem_uart_diaddr,
+                                          OSD_REG_BASE_MOD_CS, 0);
+
+    rv = osd_terminal_stop(terminal_ctx);
+    ck_assert_int_eq(rv, OSD_OK);
+
+    rv = osd_terminal_disconnect(terminal_ctx);
+    ck_assert_int_eq(rv, OSD_OK);
+
+    osd_terminal_free(&terminal_ctx);
+    ck_assert_ptr_eq(terminal_ctx, NULL);
+
+    osd_log_free(&log_ctx);
+}
+
+/**
+ * Test fixture: setup (called before each test)
+ */
+void setup(void)
+{
+    mock_hostmod_diaddr = osd_diaddr_build(1, 1);
+    mock_dem_uart_diaddr = osd_diaddr_build(target_subnet_addr, 15);
+
+    mock_host_controller_setup();
+    setup_hostmod();
+}
+
+/**
+ * Test fixture: setup (called after each test)
+ */
+void teardown(void)
+{
+    mock_host_controller_wait_for_event_tx();
+    teardown_hostmod();
+    mock_host_controller_teardown();
+}
+
+START_TEST(test_basic_init)
+{
+    setup();
+    teardown();
+}
+END_TEST
+
+// This test fails for the following chars
+// Sent - Read
+// (0x00..0x1d) - 0x1d (ASCII Group Separator, probably related to ISIG)
+// 0x7e - 0x80 (~ - non-ascii)
+// 0x7f - 0x80 (DEL - non-ascii)
+
+// If canonical mode and signals are disabled (in terminal.c)
+// 0x0a00 - 0x0d00 (CR & LF switched)
+// 0x0d00 - 0x0a00
+// 0x1100 - 0x1200
+START_TEST(test_read_from_target)
+{
+    osd_result rv;
+
+    const uint16_t TEST_START = 0x00, TEST_END = 0x10;
+
+    struct osd_packet *packet;
+    osd_packet_new(&packet, osd_packet_sizeconv_payload2data(1));
+    osd_packet_set_header(packet, mock_hostmod_diaddr, mock_dem_uart_diaddr,
+                          OSD_PACKET_TYPE_EVENT, EV_LAST);
+
+    for (uint16_t c = TEST_START; c < TEST_END; c++) {
+        packet->data.payload[0] = c;
+        rv = mock_host_controller_queue_data_packet(packet);
+        ck_assert_int_eq(rv, OSD_OK);
+    }
+
+    mock_host_controller_wait_for_event_tx();
+    osd_packet_free(&packet);
+
+    int pts = open(osd_terminal_get_pts_path(terminal_ctx), O_RDONLY | O_NOCTTY);
+    ck_assert(pts > 0);
+
+    int ret;
+    unsigned char c;
+    for (uint16_t exp = TEST_START; exp < TEST_END; exp++) {
+        ret = read(pts, &c, 1);
+        if (ret <= 0) {
+            perror("read()");
+        }
+
+        ck_assert_int_eq(c, exp);
+    }
+}
+END_TEST
+
+// This test fails for the following chars:
+// Expected - Received (Little Endian)
+// 0xa00 - 0xd00 (LF turned into CR)
+START_TEST(test_write_to_target)
+{
+    osd_result rv;
+
+    const uint16_t TEST_START = 0x00, TEST_END = 0x0100;
+    assert(TEST_END > TEST_START);
```

Comment:
Now that the whole ASCII range works with the appropriate settings, you can remove those options and just exclude whatever characters don't work (with a comment why). And as in my previous comment, use *char* here, not uint*.